### PR TITLE
trying to migrate to DOM parser

### PR DESCRIPTION
**Pull request by sudo-vaibhav**
Trying to add DOMParser so that we could target injection using CSS selectors

### CODE REVIEW

File: lib/util.js [hunk 1]
```diff
@@ -1,34 +1,34 @@
...
-  regex = createTagRegExp(tag);
-
-  return base_html.replace(regex, replacementByPriority(priority, tag, html));
-
-}
+  // this will contain the DOM representation of the template
+  const dom = new JSDOM(base_html);
+  const targetElement = dom.window.document.querySelector(location);
+  targetElement.insertAdjacentHTML(
+    priority == "high" ? "afterbegin" : "beforeend",
```
Comment (colbyfayock) [inline lib/util.js:22]: can you swap this to `===`? :)

File: package-lock.json [hunk 2]
```diff
@@ -0,0 +1,7881 @@
+{
```
Comment (colbyfayock) [inline package-lock.json:1]: i'm using yarn for this package - we dont want to include a yarn.lock in addition to a package-lock

https://classic.yarnpkg.com/lang/en/

**Comment by colbyfayock** [comment]
i think it's working  okay, i dont think you need the decoding there

the reason i had things like:

```
&#x3C;h1&#x3E;Hello world!&#x3C;/h1&#x3E;
```

was because I was having an issue getting those to run through the compile steps and output in a way that would still work in the browser as text

it looks like jsdom serializes the output a little differently than it did before, which is okay, but that means the tests will break

i would recommend checking all of the tests and seeing what the differences are, opening the compiled files locally in your browser, and make sure that they work

you can test the compilation works for the files by navigating to the example directory, the nrun:
```
../../node_modules/webpack-cli/bin/cli.js --config webpack.config.js
```

That will run the webpack compile steps, and output in the `dist` folder of that example

What i've been doing is copying the output files from dist/ into fixtures, so for instance, examples/basic/dist/index.html is basic.html

could you revert some of the changes like the single quotes? id prefer to keep those as i originally had and remain consistent with the rest of the respository

**Comment by colbyfayock** [comment]
once you get those tests fixed up we can walk through making a new test using CSS selectors :)

File: lib/util.js (hunk 1 above)
Comment (sudo-vaibhav) [reply lib/util.js:22]: will do 👍

File: package-lock.json (hunk 2 above)
Comment (sudo-vaibhav) [reply package-lock.json:1]: that's cool. I'll change it. But why do people use yarn? I never understood how that is better than npm. Node ships out of the box with npm and it has always worked for me. I'll make the amendment tho

**Comment by sudo-vaibhav** [comment]
That change of quote signs is probably because of some vs code extension or beautifier I'm using. I think its because of prettier which keeps formatting the document. It keeps changing back to the prettier formatting on every save. Is there some config file for your formatter that I can enable or use to ensure a consistent code formatting style?

**Comment by sudo-vaibhav** [comment]
The partial injection works now. Now the location tag works with selectors just like any HTML tag. And since every HTML tag is implicity a CSS selector anyways, this is backwards compatible. Updating tests and docs needs to be done though.

File: examples/analytics/dist/index.html [hunk 3]
```diff
@@ -33,19 +32,16 @@
...
-&#x3C;!-- Global site tag (gtag.js) - Google Analytics --&#x3E;
-&#x3C;script async src=&#x22;https://www.googletagmanager.com/gtag/js?id=&#x3C;%= ga_property_id %&#x3E;&#x22;&#x3E;&#x3C;/script&#x3E;
-&#x3C;script&#x3E;
+</pre><h2>partials/head.html</h2><pre>&lt;!-- Global site tag (gtag.js) - Google Analytics --&gt;
+&lt;script async src="https://www.googletagmanager.com/gtag/js?id=&lt;%= ga_property_id %&gt;"&gt;&lt;/script&gt;
+&lt;script&gt;
 window.dataLayer = window.dataLayer || [];
 function gtag(){dataLayer.push(arguments);}
-gtag(&#x27;js&#x27;, new Date());
+gtag('js', new Date());
```
Comment (colbyfayock) [inline examples/analytics/dist/index.html:43]: for what it's worth, i checked the example in the browser and these seem to be fine - i guess only the < and > need to be escaped

File: examples/analytics/dist/index.html [hunk 4]
```diff
@@ -1,9 +1,8 @@
-<!doctype html><html><head><meta charset="utf-8"><script async src="https://www.googletagmanager.com/gtag/js?id=UA-28076707-7"></script><script>window.dataLayer = window.dataLayer || [];
```
Comment (colbyfayock) [inline examples/analytics/dist/index.html:1]: we seem to have lost the meta tag here

File: examples/analytics/dist/index.html [hunk 5]
```diff
@@ -1,9 +1,8 @@
...
+<!doctype html><html><head><script async="" src="https://www.googletagmanager.com/gtag/js?id=UA-28076707-7"></script><script>window.dataLayer = window.dataLayer || [];
 function gtag(){dataLayer.push(arguments);}
 gtag('js', new Date());
 
-gtag('config', 'UA-28076707-7');</script><title>Webpack App</title><meta name="viewport" content="width=device-width,initial-scale=1"></head><body><h1>Hello world!</h1><hr/><h2>webpack.config.js</h2><pre>
-const path = require('path');
+gtag('config', 'UA-28076707-7');</script><meta charset="utf-8"><title>Webpack App</title><meta name="viewport" content="width=device-width,initial-scale=1"></head><body><h1>Hello world!</h1><hr><h2>webpack.config.js</h2><pre>const path = require('path');
```
Comment (colbyfayock) [inline examples/analytics/dist/index.html:5]: oh looks like it was just stuffed behind the script tag

this is what i was referring to in the comments from https://github.com/colbyfayock/html-webpack-partials-plugin/issues/31

i think the meta tag needs to go before any content we inject or we risk issues with how that content loads

File: examples/high-priority-head/dist/index.html [hunk 6]
```diff
@@ -1,8 +1,7 @@
-<!doctype html><html><head><meta charset="utf-8"><style>h1 {
```
Comment (colbyfayock) [inline examples/high-priority-head/dist/index.html:1]: another example of the ordering issue with the meta tag

File: lib/util.js [hunk 7]
```diff
@@ -1,73 +1,23 @@
...
  * @description takes an html string and injects new html string based on the
  *     set priority and tag location
  */
 
-module.exports.injectPartial = function(base_html, { options, html, priority, location }) {
-
-  let tag;
-  let regex;
-
-  if ( location === 'head' && priority === 'high' ) {
```
Comment (colbyfayock) [inline lib/util.js:12]: i think we'll need to re-establish some of the logic here for dynamically injecting the partials after the meta tag

File: package-lock.json (hunk 2 above)
Comment (colbyfayock) [reply package-lock.json]: good question - within the last year or so npm has caught up to a lot of features that yarn was providing that npm wasn't, such as it was generally faster and provided some API improvements. ive got used to using it for that reason and it feels a bit nicer to use, i dont have strong opinions on not changing it but dont think it should be in scope here and not worth changing unless there's a good reason to

**Comment by colbyfayock** [comment]
awesome job with all of this 💪  i think we just need to fix the meta tag placement so that we don't inadvertently cause any issues with how the page loads for people

**Comment by sudo-vaibhav** [comment]
Sorry I am slightly busy this week. So the time I can devote is slightly reduced. I already have a solution to the meta tag problem in mind. I just have to Implement it. The idea is that whenever we get a head tag with high priority, I will try to programmatically insert it after the last meta tag.

Also, It will be great if instead of just giving high and low priority options, we just give the 4 options that are there in insertAdjacentHTML by default, namly - "beforeend, afterbegin", "afterend" and "beforebegin".

File: examples/analytics/dist/index.html (hunk 3 above)
Comment (sudo-vaibhav) [reply examples/analytics/dist/index.html:43]: I don't understand exactly what you are trying to say here about escaping. Can you please elaborate. Even I am confused about the escaping and serialisation. The DOMParser might export ugly HTML code but I think it looks just fine when we look at it in the browser, which is the goal.

**Comment by colbyfayock** [comment]
now worries! there's no rush :) i appreciate you doing this

That's interesting - i think i'd be more inclined to suggest that we can add those but keep the originals available, that way it's not a breaking change for people who are already using the high and low priority

you can create an object map, so that we can maintain the original keys but map them to the right API definition

```
const map = {
  high: 'afterbegin',
  low: 'beforeend'
}

if ( map[priority] ) ...
```

**Comment by sudo-vaibhav** [comment]
I also have a question now that I have wandered around trying to solve the problems of adding partials into our webpages in my workplace. EJS seems a worthy solution to the partials injection problem, it works just fine. This brings in a question. Why do we need this plugin? Is there something I am missing here? Why would someone want to go for this plugin instead of using EJS templating which can inject partials and I think also do crazy stuff like looping and adding code dynamically?

**Comment by colbyfayock** [comment]
im not familiar with EJS - but the idea is HTML Webpack Plugin is a popular way to build out sites with Webpack. This extends it making it easier to manage small snippets of code throughout the generated HTML, rather than having to manage the index.html file as a template

i dont think this is really doing anything crazy

**Comment by colbyfayock** [comment]
if you dont have any interest in continuing on this - i could branch off of your work and wrap it up. i appreciate what you did either way :)

**Comment by sudo-vaibhav** [comment]
Don't worry, I'll keep working on this. Its just that right now, the code at my place is turning into a weird sphagetti of ejs partials and html partials due to the lack of precision in injection in this plugin. I am using ejs wherever this plugin doesn't fulfill our needs. And also one weird thing , Even with the new code I wrote, the injection doesn't work when I test it in my workplace project. Even though examples work fine in the plugin , in the real projects, whatever I wrote is not working and I can't figure out why and this is frustrating me. I agree that the meta tag part is not finished in that code but that shouldn't matter in my use case as I am not injecting anything in head anyway. I'm confused why my new code addition to this plugin doesn't work in production if it works on the examples, in my workplace project code, the plugin behaves as if it isn't even there. No injections happen. I'm confused 🤔

**Comment by colbyfayock** [comment]
do you see any errors in the terminal? or maybe are you on an older version of webpack?

**Comment by sudo-vaibhav** [comment]
Also could the failure to inject partials in my work project code be happening because I'm always giving multiple target html files to insert into? Like in all the examples, either we dont specify a base index.html at all or if we do, its just one file. But in my project, I'm always passing an array of multiple files which is cool according to the docs and also works on older versions of this plugin. Did I break something? Can we make an example for multiple files being targeted by the same partial?

**Comment by sudo-vaibhav** [comment]
I am doing crazy partial injection in the real project which is much more complex than the examples, I'll try to show it to you in a gist. Its real bonkers. I'm putting multiple partials in one file and also putting the same partial in multiple files.

**Comment by sudo-vaibhav** [comment]
could you maybe point me to the part of the code which handles what happens if the same partial gets multiple files to target? or is it something that's handled by some webpack thingy outside of our control?

**Comment by sudo-vaibhav** [comment]
Here is a gist of my webpack config: https://gist.github.com/sudo-vaibhav/836e19d6a4f747a7a483d4804819743a
And my new code is not working at all for this config... Tell me if that's not the most insane use for your plugin you have ever seen 😄

File: examples/analytics/dist/index.html (hunk 3 above)
Comment (colbyfayock) [reply examples/analytics/dist/index.html:43]: yeah i think we're good :) im just talking out loud

**Comment by colbyfayock** [comment]
haha ill definitely say it's an impressive use!

I'm totally up for you creating any sort of example that helps. for that - i would actually recommend creating a separate branch, then create the example there, which we can merge in with a different pull request

once that's working with the original plugin code with passing tests, you can merge that branch into this branch, where we might be able to see where it's failing

does that make sense?

as for the injecting to multiple templates - here's the code: https://github.com/colbyfayock/html-webpack-partials-plugin/blob/061c1eb3877e1fa77a7404512bef2db9bd54d3ab/index.js#L29 

`template_filename` isn't actually utilized to perform an action, rather it's used to identify the template that this plugin should inject to

the logic in that file should probably be simplified a bit to make it more readable, as it's kinda hard to read in that big return statement

so for example, in this example: https://github.com/colbyfayock/html-webpack-partials-plugin/blob/688d8e0ff372979881dc2f34266dbf3aee6bc891/examples/page-template/webpack.config.js

in the HTMLWebpackPlugin configuration, we have a template that we're defining as `about.html`. in one of the instances of HTMLWebpackPartialsPlugin, we're telling the plugin that we want to target that `about.html` template

**Comment by colbyfayock** [comment]
hey @sudo-vaibhav i know its been a little while, but do you still wanna wrap this up? some good work in here! and i think some others would love the feature too

**Comment by colbyfayock** [comment]
closing this for now - feel free to reopen if you come back to it and catch the branch up

**Pull request closed without merging**